### PR TITLE
eth: fixed chaindb upgrade

### PR DESCRIPTION
**Pull request by zsfelfoldi**

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Jul 19 12:05:17 UTC 2016_

File: eth/db_upgrade.go [hunk 1]
```diff
@@ -93,13 +93,19 @@ func upgradeSequentialKeys(db ethdb.Database) (stopFn func()) {
 func upgradeSequentialCanonicalNumbers(db ethdb.Database, stopFn func() bool) (error, bool) {
 	prefix := []byte("block-num-")
 	it := db.(*ethdb.LDBDatabase).NewIterator()
+	defer func() {
+		it.Release()
+	}()
```
Comment (fjl) [inline eth/db_upgrade.go]: `defer it.Release()`

Comment (fjl) [inline eth/db_upgrade.go]: Or no, it's actually correct this way here because the value of `it` changes in the loop.

File: eth/db_upgrade.go [hunk 2]
```diff
@@ -175,6 +187,9 @@ func upgradeSequentialBlocks(db ethdb.Database, stopFn func() bool) (error, bool
 func upgradeSequentialOrphanedReceipts(db ethdb.Database, stopFn func() bool) (error, bool) {
 	prefix := []byte("receipts-block-")
 	it := db.(*ethdb.LDBDatabase).NewIterator()
+	defer func() {
+		it.Release()
+	}()
```
Comment (fjl) [inline eth/db_upgrade.go]: But you can write `defer it.Release()` here.

Comment (zsfelfoldi) [inline eth/db_upgrade.go]: done

**Comment by fjl** [comment]
Code change :+1:. I'm worried that swapping out the iterator might cause consistency issues, but I guess we'll see the issue when it happens.

**Comment by karalabe** [comment]
@fjl It cannot really because old entries are never added. Updated code will always create new-prefixed entries, so old ones can gradually phase out.

**Comment by karalabe** [comment]
I'll do a fast sync and try to upgrade afterwards, if all's ok, we can merge this.

**Comment by karalabe** [comment]
The bug was fixed now. LGTM :+1: 

Btw, there's still something we should investigate: an updated database uses 1GB more disk space than the original one.

**Pull request merged**